---
title: "The vaginal microbiome and preterm birth"
author: "Hardik I Parikh"
date: "03/14/2019"
output: html_document
---

## Extended Data Figure 5:  

Taxanomic 16S rRNA, metagenomic and metatranscriptomic pathway analysis

```{r setup, include=FALSE}
library(tidyverse)
library(reshape2)
library(vegan)
library(dendsort)
library(ggpubr)
library(cowplot)
library(grid)
library(gridExtra)
library(RColorBrewer)
set.seed(12345)
```


#### Load Data

```{r, echo=TRUE}
# ptb45 okpt
ptb45.okpt <- read_delim("samples_list.txt", delim = "\t")
  
# read mefit stats for min 1000hq reads
mefit <- read_delim("mefit_stats.txt", delim = "\t") %>%
  filter(HQReads > 1000)

# read stirrups v2 profiles
ssdata <- read_delim("stirrups_profiles.txt", delim = "\t") %>%
  filter(
         SampleID %in% mefit$SampleID &
         ATprop > 0.01)

# re-calculate proportions to 100.0
tmp <- ssdata %>%
  group_by(SampleID) %>%
  summarise(TotalReads = sum(No_of_Reads))
ssdata <- left_join(ssdata, tmp, by="SampleID") %>%
  mutate(RelAb = (No_of_Reads/TotalReads)*100)
#foo <- aggregate(ssdata$RelAb, by=list(SampleID=ssdata$SampleID), FUN=sum)

# term/preterm subjects
termstatus <- read_delim("metadata.txt", delim = "\t") %>%
  mutate(preg = ifelse(Group == "control", "term", Group))

# read custom colors
taxacolors <- read_delim("taxa_colors.txt", delim = "\t")
vtcolors <- read_delim("vt_colors.txt", delim = "\t")

# wmgs read Counts
wmgsreads <- read_delim("wmgs_read_counts.txt", delim="\t")
# wmgs path abundances
wmgspath <- read_delim("wmgs_humann2_pathAbundance.txt", delim = "\t") %>%
  filter(!(Pathway %in%  c("UNMAPPED", "UNINTEGRATED"))) 
  

# wmts read Counts
# wmtsreads 
wmtsreads <- read_delim("wmts_read_counts.txt", delim="\t")
# wmts path abundances
wmtspath <- read_delim("wmts_humann2_pathAbundance.txt", delim = "\t") %>%
  filter(!(Pathway %in%  c("UNMAPPED", "UNINTEGRATED"))) %>%
  mutate(SampleID_MV1R = SampleID) %>%
  rowwise() %>%
  mutate(SampleID = gsub("MV1R", "MV1D", SampleID))

```

#### Assign Vagitypes to each Sample

```{r, echo=TRUE}

### get vagitypes for each sample
### Vagitype definition: >= 30% of dominant taxa
getVagitype <- function(propData) {
  ## identify vagitypes
  vagTypes <- apply(propData, 1, which.max)
  maxProp <- propData[matrix(c(1:nrow(propData),vagTypes), ncol=2)]
  vagTypes <- colnames(propData)[vagTypes]
  vagTypes[maxProp < 30.0] <- "No Type"
  
  ## my vagTypeDF
  vagTypeDF <- data.frame(cbind(rownames(propData), vagTypes))
  vagTypeDF$vagTypes.2 <- NA
  for(i in 1:nrow(vagTypeDF)) {
    if(grepl("crispatus", vagTypeDF[i,2])) {vagTypeDF[i,3] <- "L.crispatus"}
    else if(grepl("iners", vagTypeDF[i,2])) {vagTypeDF[i,3] <- "L.iners"}
    else if(grepl("jensenii", vagTypeDF[i,2])) {vagTypeDF[i,3] <- "L.jensenii"}
    else if(grepl("gasseri", vagTypeDF[i,2])) {vagTypeDF[i,3] <- "L.gasseri"}
    else if(grepl("BVAB1", vagTypeDF[i,2])) {vagTypeDF[i,3] <- "BVAB1"}
    else if(grepl("vaginalis", vagTypeDF[i,2])) {vagTypeDF[i,3] <- "G.vaginalis"}
    else if(grepl("No Type", vagTypeDF[i,2])) {vagTypeDF[i,3] <- "NoType"}
    else {vagTypeDF[i,3] <- "Other"}
  }
  
  ## set colnames
  colnames(vagTypeDF) <- c("SampleID", "vagitype", "vagitype.2") 
  
  return(vagTypeDF)
}

propData <- acast(ssdata, SampleID~Taxa, value.var = "RelAb", fill=0.0)
vagitypeDF <- getVagitype(propData)

ssdata <- left_join(ssdata, vagitypeDF, by="SampleID")
```

#### Add pregnancy status, trimester,  custom colors

```{r, echo=TRUE}
# Add pregnancy status 
ssdata <- ssdata %>%
  rowwise() %>%
  mutate(ParticipantID = strsplit(as.character(SampleID), "_")[[1]][1]) %>%
  mutate(KitID = paste(strsplit(as.character(SampleID), "_")[[1]][1], strsplit(as.character(SampleID), "_")[[1]][2], sep="_"))
termstatus <- termstatus %>%
  filter(ParticipantID %in% ssdata$ParticipantID) %>%
  select(ParticipantID, preg) %>%
  unique()
ssdata <- left_join(ssdata, termstatus, by="ParticipantID") %>%
  mutate(preg = factor(preg, levels = c("term", "preterm")))

# Add custom colors
ssdata <- left_join(ssdata, taxacolors, by="Taxa") %>%
  mutate(TaxaColor = ifelse(is.na(TaxaColor), "#faf0e6", TaxaColor)) 

ssdata <- left_join(ssdata, vtcolors, by="vagitype")

```

#### Sort Samples by Vagitype

```{r, echo=TRUE}
uniq.vagitypes <- ssdata %>%
  select(vagitype.2) %>%
  unique()

# set default vagitype order
myVTOrder <- c("L.crispatus", "L.iners", "L.jensenii", "L.gasseri", "BVAB1", "G.vaginalis", "Other", "NoType")

# get sample order
sampleOrder <- vector(mode="character", length=0)

for (x in myVTOrder) {
  if (x %in% uniq.vagitypes$vagitype.2) {
    ssdata_subset <- ssdata %>%
      filter(vagitype.2 == x)
    if (length(unique(ssdata_subset$SampleID)) > 1) {
      propData.subset <- acast(ssdata_subset, SampleID~Taxa, value.var = "RelAb", fill=0.0)
      bcdist <- vegdist(propData.subset, method="bray")
      hclustBC <- hclust(bcdist, method="ward.D2")
      dd <- dendsort(as.dendrogram(hclustBC))
      hclustBC.order <- as.hclust(dd)
      sampleOrder <- c(sampleOrder, hclustBC.order$labels[c(hclustBC.order$order)])
    } 
    else {
      sampleOrder <- c(sampleOrder, as.character(unique(ssdata_subset$SampleID)))
    }
  }
}

# set sample order
ssdata <- ssdata %>%
  mutate(SampleID = factor(SampleID, levels = sampleOrder)) 
ssdata$TaxaColor <- as.factor(ssdata$TaxaColor)
```

#### Subset for samples that pass seq depths

16S reads > 1000  
WMGS reads > 100,000  
WMTS reads > 100,000

```{r, echo=TRUE}
wmgs_hq <- wmgsreads %>% 
  filter(InputMetagenome > 100000) %>%
  rowwise() %>%
  mutate(KitID = paste(strsplit(as.character(SampleID), "_")[[1]][1], strsplit(as.character(SampleID), "_")[[1]][2], sep="_"))

wmts_hq <- wmtsreads %>%
  filter(InputMetatranscriptome > 100000) %>%
  rowwise() %>%
  mutate(KitID = paste(strsplit(as.character(SampleID), "_")[[1]][1], strsplit(as.character(SampleID), "_")[[1]][2], sep="_"))

hq_kits <- intersect(intersect(ssdata$KitID, wmgs_hq$KitID), wmts_hq$KitID)

```


#### Stacked Bars 
```{r, echo=TRUE}
myStackedBarPlot <- function(inputDF) {
  p <- ggplot(inputDF, aes(SampleID, RelAb, fill=TaxaColor, group=RelAb)) + 
                geom_bar(stat = 'identity', position='stack', width = 1) +
                scale_fill_manual(values=levels(inputDF$TaxaColor)) +
                labs(x="Sample", y="Taxa Abundance") + 
                facet_grid(~preg, scales = "free_x") +
                theme_bw() + 
                ggtitle("") + 
                labs(x="Samples", y="Relative Abundance") + 
                theme(strip.text=element_blank(),
                      axis.text.y=element_text(face="bold",size=10),
                      axis.text.x = element_blank(),
                      axis.ticks.x=element_blank(),
                      axis.title = element_text(size=12, face="bold"),
                      plot.title = element_text(size=14, face="bold"),
                      legend.position = "none") 
  return(p)
}

ssdata_hq <- ssdata %>%
  filter(KitID %in% hq_kits)
sb <- myStackedBarPlot(ssdata_hq)
sb
```

```{r, echo=TRUE, fig.width=14, fig.height=3.5}

### Plot Legend
# Only show selected taxa in legend
myTaxaList <- c("Lactobacillus_crispatus_cluster", "Lactobacillus_iners", 
                "Lactobacillus_gasseri_cluster", "Lactobacillus_jensenii", 
                "Lachnospiraceae_BVAB1", "Gardnerella_vaginalis",
                "Atopobium_vaginae", "Sneathia_amnii", 
                "Lactobacillus_delbrueckii", "Prevotella_cluster2", 
                "Prevotella_bivia", 
                "Megasphaera_OTU70_type1", "Megasphaera_OTU71_type2", 
                "Clostridiales_BVAB2", 
                "Mycoplasma_hominis", "Streptococcus_cluster29", 
                "Streptococcus_agalactiae", "Mycoplasma_girerdii", 
                "Dialister_micraerophilus", "Parvimonas_OTU142",
                "Ureaplasma_cluster23", "TM7_OTU-H1" , "Aerococcus_christensenii" )
plotDF <- ssdata %>%
  filter(Taxa %in% myTaxaList) %>%
  rbind(c("EP407012_K20_MV1D","Other Taxa", 0, 0.00, 63265, 0.00, "Gardnerella_vaginalis", "G.vaginalis", "EP407012", "EP407012_K20", "term", "#faf0e6", "#d62728", "red"))  #add row for Other Taxa

tmp.plot <- ggplot(plotDF, aes(SampleID, RelAb, fill=TaxaColor)) + 
  geom_bar(stat = "identity", position = "stack") + 
  theme(legend.position = "bottom", 
        legend.text = element_text(size=10),
        legend.text.align = 0,
        legend.title = element_blank(), 
        legend.justification = "center",
        plot.margin = unit(c(0, 0, 0, 0), "cm")) + 
  scale_fill_manual(values=c("#fff5aa", "#aec7e8", "#c5b0d5", 
                            "#ffbb78", "#ff7f0e", "#d62728",
                             "#c49c94", "#9467bd", "#ff9896",
                             "#1f77b4", "#17becf", "#f7b6d2", "gray60", 
                            "#dee0e5", "#98df8a", 
                             "#9edae5", "#67a9cf", "#7f7f7f", 
                             "#2ca02c", "#dbdb8d", "#bcbd22", 
                            "#e377c2", "#a5acaf", "#faf0e6"), 
                    labels=c(expression(italic("Lactobacillus crispatus")), 
                             expression(italic("Lactobacillus iners")), 
                             expression(italic("Lactobacillus gasseri")),
                             expression(italic("Lactobacillus jensenii")),
                             expression(paste(italic('"Lachnospiraceae"'), plain(" BVAB1"))), 
                             expression(italic("Gardnerella vaginalis")),
                             expression(italic("Atopobium vaginae")), 
                             expression(italic("Sneathia amnii")),
                             expression(italic("Lactobacillus delbrueckii")),
                             expression(paste(italic("Prevotella"), plain(" cluster2"))), 
                             expression(italic("Prevotella bivia")),
                             expression(paste(italic("Megasphaera"), plain(" OTU70 sp. Type1"))),
                             expression(paste(italic("Megasphaera"), plain(" OTU71 sp. Type2"))),
                             expression(paste(italic("Clostridiales"), plain(" BVAB2"))),
                             expression(italic("Mycoplasma hominis")),
                             expression(paste(italic("Streptococcus"), plain(" cluster29"))),
                             expression(italic("Streptococcus agalactiae")), 
                             expression(italic("Mycoplasma girerdii")),
                             expression(italic("Dialister micraerophilus")),
                             expression(paste(italic("Parvimonas"), plain(" OTU142"))),
                             expression(paste(italic("Ureaplasma"), plain(" cluster23"))),
                             expression(plain("TM7 OTU-H1")),
                             expression(italic("Aerococcus_christensenii")),
                             "Other Taxa")) + 
  guides(fill=guide_legend(nrow = 6))
legend <- get_legend(tmp.plot)
legend$grobs[[1]]$widths[5] <- unit(0.25, "cm")
legend$grobs[[1]]$widths[6] <- unit(4.75, "cm")
legend$grobs[[1]]$widths[9] <- unit(0.25, "cm")
legend$grobs[[1]]$widths[10] <- unit(5.25, "cm")
legend$grobs[[1]]$widths[13] <- unit(0.25, "cm")
legend$grobs[[1]]$widths[14] <- unit(5.25, "cm")
legend$grobs[[1]]$widths[17] <- unit(0.25, "cm")
legend$grobs[[1]]$widths[18] <- unit(4.75, "cm")
tlegend <- ggdraw(legend)
tlegend
```

```{r}

#ExtDataFig8A
ggsave("ExtDataFigure5a_16S.eps", sb, units = "in", width = 10, height = 3, dpi=600)
ggsave("ExtDataFigure5a_16Slegend.eps", tlegend, units = "in", width = 10, height = 1.5, dpi=600)
```

#### WMGS Pathway Abundances

```{r, echo=TRUE, fig.width=6, fig.height=6}
wmgspath_tot <- wmgspath %>%
  group_by(SampleID) %>%
  summarise(totRelAb = sum(`Normalized-RelativeAbundance`))

wmgspath_hq <- wmgspath %>%
  rowwise() %>%
  mutate(ParticipantID = strsplit(as.character(SampleID), "_")[[1]][1]) %>%
  mutate(KitID = paste(strsplit(as.character(SampleID), "_")[[1]][1], strsplit(as.character(SampleID), "_")[[1]][2], sep="_")) %>% 
  filter(KitID %in% hq_kits) %>%
  left_join(., wmgspath_tot, by="SampleID") %>%
  mutate(NormRelAb = `Normalized-RelativeAbundance`/totRelAb) %>%
  left_join(., termstatus, by="ParticipantID") %>%
  mutate(SampleID = factor(SampleID, levels = levels(ssdata$SampleID))) %>%
  mutate(preg = factor(preg, levels = c("term", "preterm")))
  
mycols <- colorRampPalette(brewer.pal(12, "Paired"))(366)

wmgs_allpath <- ggplot(wmgspath_hq, aes(SampleID, NormRelAb, fill=Pathway)) + 
                  geom_bar(stat = 'identity', position='stack', width = 1) +
                  labs(x="Sample", y="Relative Abundance") + 
                  scale_fill_manual(values = mycols) + 
                  facet_grid(~preg, scales = "free_x") +
                  theme_bw() + 
                  ggtitle("WMGS: Pathway Abundances") + 
                  labs(x="Samples", y="Pathway Abundance") + 
                  theme(strip.text=element_text(face="bold",size=12),
                        axis.text.y=element_text(face="bold",size=10),
                        axis.text.x = element_blank(),
                        axis.ticks.x=element_blank(),
                        axis.title = element_text(size=12, face="bold"),
                        plot.title = element_text(size=14, face="bold"),
                        legend.position = "none") 
wmgs_allpath
```


#### WMTS Pathway Abundances
```{r, echo=TRUE}
wmtspath_tot <- wmtspath %>%
  group_by(SampleID) %>%
  summarise(totRelAb = sum(`Normalized-RelativeAbundance`))

wmtspath_hq <- wmtspath %>%
  rowwise() %>%
  mutate(ParticipantID = strsplit(as.character(SampleID), "_")[[1]][1]) %>%
  mutate(KitID = paste(strsplit(as.character(SampleID), "_")[[1]][1], strsplit(as.character(SampleID), "_")[[1]][2], sep="_")) %>% 
  filter(KitID %in% hq_kits) %>%
  left_join(., wmtspath_tot, by="SampleID") %>%
  mutate(NormRelAb = `Normalized-RelativeAbundance`/totRelAb) %>%
  left_join(., termstatus, by="ParticipantID") %>%
  mutate(SampleID = factor(SampleID, levels = levels(ssdata$SampleID))) %>%
  mutate(preg = factor(preg, levels = c("term", "preterm")))

mycols <- colorRampPalette(brewer.pal(12, "Paired"))(366)

wmts_allpath <- ggplot(wmtspath_hq, aes(SampleID, NormRelAb, fill=Pathway)) + 
                  geom_bar(stat = 'identity', position='stack', width = 1) +
                  labs(x="Sample", y="Relative Abundance") + 
                  scale_fill_manual(values = mycols) + 
                  facet_grid(~preg, scales = "free_x") +
                  theme_bw() + 
                  ggtitle("WMTS: Pathway Abundances") + 
                  labs(x="Samples", y="Pathway Abundance") + 
                  theme(strip.text=element_text(face="bold",size=12),
                        axis.text.y=element_text(face="bold",size=10),
                        axis.text.x = element_blank(),
                        axis.ticks.x=element_blank(),
                        axis.title = element_text(size=12, face="bold"),
                        plot.title = element_text(size=14, face="bold"),
                        legend.position = "none") 

wmts_allpath
```


#### WMGS Pathways: Selected Pathways

```{r, echo=TRUE}
myPaths <- c("PWY-7219__adenosine_ribonucleotides_de_novo_biosynthesis", 
             "UDPNAGSYN-PWY__UDP-N-acetyl-D-glucosamine_biosynthesis_I", 
             "PWY-5100__pyruvate_fermentation_to_acetate_and_lactate_II",
             "NONOXIPENT-PWY__pentose_phosphate_pathway_(non-oxidative_branch)",
             "PWY-7229__superpathway_of_adenosine_nucleotides_de_novo_biosynthesis_I",
             "ANAGLYCOLYSIS-PWY__glycolysis_III_(from_glucose)",
             "PWY-6387__UDP-N-acetylmuramoyl-pentapeptide_biosynthesis_I_(meso-diaminopimelate_containing)",
             "PEPTIDOGLYCANSYN-PWY__peptidoglycan_biosynthesis_I_(meso-diaminopimelate_containing)",
             "PWY-5686__UMP_biosynthesis",
             "COA-PWY__coenzyme_A_biosynthesis_I")

wmgspath_t10 <- wmgspath_hq %>%
                  filter(Pathway %in% myPaths)

mycols2 <- brewer.pal(10, "Spectral")
pathlabs <- c("glycolysis_III_(from_glucose)", 
              "coenzyme_A_biosynthesis_I",
              "pentose_phosphate_pathway_(non-oxidative_branch)",
              "peptidoglycan_biosynthesis_I_(meso-diaminopimelate_containing)",
              "pyruvate_fermentation_to_acetate_and_lactate_II",
              "UMP_biosynthesis",
              "UDP-N-acetylmuramoyl-pentapeptide_biosynthesis_I_(meso-diaminopimelate_containing)",
              "adenosine_ribonucleotides_de_novo_biosynthesis",
              "superpathway_of_adenosine_nucleotides_de_novo_biosynthesis_I",
              "UDP-N-acetyl-D-glucosamine_biosynthesis_I")
myplot <- ggplot(wmgspath_t10, aes(SampleID, NormRelAb, fill=Pathway)) + 
              geom_bar(stat = 'identity', position='stack', width = 1) +
              labs(x="Sample", y="Relative Abundance") + 
              scale_fill_manual(labels = pathlabs, values = mycols2) + 
              facet_grid(~preg, scales = "free_x") +
              theme_bw() + 
              ggtitle("") + 
              labs(x="Samples", y="Pathway Abundance") + 
              guides(fill=guide_legend(ncol = 2, title = "", title.position = "top")) +
              ylim(0, 1) + 
              theme(strip.text=element_blank(),
                    axis.text.y=element_text(face="bold",size=10),
                    axis.text.x = element_blank(),
                    axis.ticks.x=element_blank(),
                    axis.title = element_text(size=12, face="bold"),
                    plot.title = element_text(size=14, face="bold"),
                    legend.text = element_text(size=10),
                    legend.title = element_text(size=14, face="bold"))

wmgs_somepath <- myplot + 
  theme(legend.position = "none")
#wmgs_somepath
#ggdraw(plegend)
```


#### WMTS Pathways: Selected Pathways

```{r, echo=TRUE}
wmts_somepath <- wmtspath_hq %>%
                  filter(Pathway %in% myPaths) %>%
                  ggplot(., aes(SampleID, NormRelAb, fill=Pathway)) + 
                    geom_bar(stat = 'identity', position='stack', width = 1) +
                    labs(x="Sample", y="Relative Abundance") + 
                    scale_fill_manual(values = mycols2) + 
                    facet_grid(~preg, scales = "free_x") +
                    theme_bw() + 
                    ggtitle("") + 
                    labs(x="Samples", y="Pathway Abundance") + 
                    ylim(0, 1) + 
                    theme(strip.text=element_blank(),
                          axis.text.y=element_text(face="bold",size=10),
                          axis.text.x = element_blank(),
                          axis.ticks.x=element_blank(),
                          axis.title = element_text(size=12, face="bold"),
                          plot.title = element_text(size=14, face="bold"),
                          legend.position = "none") 
#wmts_somepath
```

#### Pathway legend
```{r, echo=TRUE}
plegend <- get_legend(myplot)
ggdraw(plegend)
```
```{r}
ggsave("ExtDataFigure8b_WMGS.eps", wmgs_allpath, width = 10, height = 3, units = "in", dpi=600)
ggsave("ExtDataFigure8b_WMTS.eps", wmts_allpath, width = 10, height = 3, units = "in", dpi=600)
ggsave("ExtDataFigure8c_WMGS.eps", wmgs_somepath, width = 10, height = 3, units = "in", dpi=600)
ggsave("ExtDataFigure8c_WMTS.eps", wmts_somepath, width = 10, height = 3, units = "in", dpi=600)
ggsave("ExtDataFigure8_legend.eps", plegend, width = 12, height = 2, units = "in", dpi=600)

```
